### PR TITLE
[EOS-19582][R2 MD DI] R2 DI: Validate metadata consistency/integrity

### PR DESCRIPTION
**Pull request by dsurnin**
-----
[View rendered docs/design/object-protection.md](https://github.com/dsurnin/cortx-s3server/blob/EOS-19582-R2-DI/docs/design/object-protection.md)
[View rendered docs/object-protection-testing.md](https://github.com/dsurnin/cortx-s3server/blob/EOS-19582-R2-DI/docs/object-protection-testing.md)

### CODE REVIEW

**Comment by dsurnin** [comment]
rerun premerge job
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/1027/

File: dev-starts3.sh [hunk 1]
```diff
@@ -134,19 +134,6 @@ ulimit -c unlimited
 #Set max open file limit to 10240
 ulimit -n 10240
 
-# Run m0dixinit
```
Comment (knrajnambiar76) [inline dev-starts3.sh:137]: Not needed?

File: s3config.release.yaml.sample [hunk 2]
```diff
@@ -19,6 +19,8 @@
 
 S3Config_Sections: [S3_SERVER_CONFIG, S3_AUTH_CONFIG, S3_MOTR_CONFIG, S3_THIRDPARTY_CONFIG, S3_VERSION_CONFIG]
 S3_SERVER_CONFIG:                                       # Section for S3 Server
+   S3_DI_DISABLE_DATA_CORRUPTION_IEM: false             # disable IEM in case of data integrity checksum fails
```
Comment (knrajnambiar76) [inline s3config.release.yaml.sample:22]: disable sending of IEM messages by s3server in case ....

File: server/s3_object_metadata.cc [hunk 3]
```diff
@@ -364,20 +414,35 @@ void S3ObjectMetadata::load_successful() {
...
-    s3_log(S3_LOG_WARN, request_id, "Object metadata load failed\n");
-    state = S3ObjectMetadataState::failed_to_launch;
-  } else {
-    s3_log(S3_LOG_WARN, request_id, "Object metadata load failed\n");
-    state = S3ObjectMetadataState::failed;
+  switch (motr_kv_reader->get_state()) {
+    case S3MotrKVSReaderOpState::failed_to_launch:
+      state = S3ObjectMetadataState::failed_to_launch;
+      s3_log(S3_LOG_WARN, request_id,
+             "Object metadata load failed to launch - ServiceUnavailable");
```
Comment (knrajnambiar76) [inline server/s3_object_metadata.cc:421]: Do ensure that in latest logging (Anatoliy's changes) \n is put in at end

File: st/clitests/aws_iam_credential_file [hunk 4]
```diff
@@ -1,5 +1,3 @@
 [default]
 aws_access_key_id = G6OFkXkYTyS_YRmxenLbBg
-aws_secret_access_key = KPykpvZqBQFysZ8vkmGwrPLXyafnOJ52n/qpRVOF
-
-
+aws_secret_access_key = KPykpvZqBQFysZ8vkmGwrPLXyafnOJ52n/qpRVOF
```
Comment (knrajnambiar76) [inline st/clitests/aws_iam_credential_file:3]: This file not needed to be checked in

**Review by knrajnambiar76 — APPROVED** [review]

File: dev-starts3.sh (hunk 1 above)
Comment (dsurnin) [reply dev-starts3.sh:137]: yes, it was removed after discussion with motr team

File: s3config.release.yaml.sample (hunk 2 above)
Comment (dsurnin) [reply s3config.release.yaml.sample]: fixed

File: server/s3_object_metadata.cc (hunk 3 above)
Comment (dsurnin) [reply server/s3_object_metadata.cc]: fixed

File: st/clitests/aws_iam_credential_file (hunk 4 above)
Comment (dsurnin) [reply st/clitests/aws_iam_credential_file]: fixed

File: st/clitests/multipart_md_integrity.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
+{"desc": "Default system test plan for Metada Integrity check for multipart with auto bucket name gen and auto key name gen",
```
Comment (t7ko-seagate) [inline st/clitests/multipart_md_integrity.json:1]: Metada -> Metadata

File: st/clitests/regular_md_integrity.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
+{"desc": "Default system test plan for Metada Integrity check",
```
Comment (t7ko-seagate) [inline st/clitests/regular_md_integrity.json:1]: Metada -> Metadata

File: server/s3_iem.h [hunk 7]
```diff
@@ -56,19 +56,6 @@ extern S3Option* g_option_instance;
...
-  do {                                                               \
-    if (loglevel == LOG_INFO) {                                      \
-      s3_syslog(loglevel, "IEC:IS" event_code ":" event_description, \
-                ##__VA_ARGS__);                                      \
-    } else {                                                         \
-      s3_syslog(LOG_ERR, "IEC:ES" event_code ":" event_description,  \
-                ##__VA_ARGS__);                                      \
-    }                                                                \
-  } while (0)
-
```
Comment (t7ko-seagate) [inline server/s3_iem.h:71]: is it intended that we remove this macro? not a misprint?

**Review by t7ko-seagate — APPROVED** [review]

File: st/clitests/regular_md_integrity.json (hunk 6 above)
Comment (dsurnin) [reply st/clitests/regular_md_integrity.json]: fixed

File: st/clitests/multipart_md_integrity.json (hunk 5 above)
Comment (dsurnin) [reply st/clitests/multipart_md_integrity.json]: fixed

File: server/s3_iem.h (hunk 7 above)
Comment (dsurnin) [reply server/s3_iem.h:71]: this macro were used nowhere, its functionality fully duplicated in the neighbor macro which is used, no additional comments on why it should be there. so I deleted it

**Comment by dsurnin** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/1071/

File: docs/design/object-protection.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+**TBD**
+
+
+## IEM alert details
+
+### Alert ID: 0030060004
+
+Current version of alert text:
+
+> Data integrity validation failure (content checksum mismatch).
```
Comment (nileshgovande) [inline docs/design/object-protection.md:30]: Please rewrite as R2 DI

**Comment by dsurnin** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/1079/

File: st/clitests/integrity.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+
```
Comment (nileshgovande) [inline st/clitests/integrity.py:2]: Need documentation

File: st/clitests/md_integrity.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
+#!/usr/bin/env python3
+
+import json
```
Comment (nileshgovande) [inline st/clitests/md_integrity.py:3]: Need documentation and file header

File: docs/object-protection-testing.md [hunk 11]
```diff
@@ -0,0 +1,428 @@
+### License
+
```
Comment (nileshgovande) [inline docs/object-protection-testing.md:2]: Pls document here that data  and metadata FI is still WIP

File: st/clitests/integrity.py (hunk 9 above)
Comment (dsurnin) [reply st/clitests/integrity.py:2]: added

File: st/clitests/md_integrity.py (hunk 10 above)
Comment (dsurnin) [reply st/clitests/md_integrity.py:24]: added

File: docs/object-protection-testing.md (hunk 11 above)
Comment (dsurnin) [reply docs/object-protection-testing.md:2]: added

File: docs/design/object-protection.md (hunk 8 above)
Comment (dsurnin) [reply docs/design/object-protection.md:32]: added

**Comment by dsurnin** [comment]
http://eos-jenkins.mero.colo.seagate.com/job/Cortx-Automation/job/S3Server/job/S3-Compatibility-Test-Ceph/173/
http://eos-jenkins.mero.colo.seagate.com/job/Cortx-Automation/job/S3Server/job/S3-Compatibility-Test-Splunk/166/

**Pull request merged**